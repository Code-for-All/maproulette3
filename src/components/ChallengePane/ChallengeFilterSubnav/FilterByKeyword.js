import React, { Component } from 'react'
import PropTypes from 'prop-types'
import _keys from 'lodash/keys'
import _without from 'lodash/without'
import _isEmpty from 'lodash/isEmpty'
import _first from 'lodash/first'
import { injectIntl } from 'react-intl'
import { CHALLENGE_CATEGORY_OTHER,
         categoryMatchingKeywords,
         combinedCategoryKeywords,
         keywordLabels }
       from '../../../services/Challenge/ChallengeKeywords/ChallengeKeywords'
import NavDropdown from '../../Bulma/NavDropdown'
import MenuList from '../../Bulma/MenuList'
import OtherKeywordsOption from './OtherKeywordsOption'
import messages from './Messages'

/**
 * FilterByKeyword displays a nav dropdown containing options for filtering
 * challenges by category keyword or a manually-entered keyword. The redux
 * store is updated to reflect the chosen category or keyword.
 *
 * @see See ChallengeCategoryKeywords
 *
 * @author [Neil Rotstan](https://github.com/nrotstan)
 */
export class FilterByKeyword extends Component {
  /**
   * Update the challenge filter with the selected value
   *
   * @private
   */
  updateFilter = ({ value }) => {
    if (value === null) {
      this.props.removeChallengeFilters(['keywords'])
    }
    else {
      this.props.setKeywordFilter(combinedCategoryKeywords[value])
    }
  }

  setOtherKeywords = keywordString => {
    this.props.setKeywordFilter([keywordString])
  }

  clearOtherKeywords = () => {
    this.updateFilter({value: null})
  }

  render() {
    const localizedKeywordLabels = keywordLabels(this.props.intl, true)

    const categories = _without(_keys(combinedCategoryKeywords), 'other')
    const activeCategory = categoryMatchingKeywords(this.props.challengeFilter.keywords, true)
    const selectOptions = categories.map(keyword => ({
      key: keyword,
      text: localizedKeywordLabels[keyword],
      value: keyword,
    }))

    // Add 'Anything' option to start of dropdown
    const anyOption = {
      key: 'any',
      text: localizedKeywordLabels.any,
      value: null,
    }
    selectOptions.unshift(anyOption)

    // If the active category doesn't match a known category, then it's a
    // manually entered ("other") keyword
    const otherKeyword = activeCategory === CHALLENGE_CATEGORY_OTHER ?
                         _first(this.props.challengeFilter.keywords) : null

    // Add 'other' box for manually entering other keywords not included in menu.
    selectOptions.push({
      Renderable: OtherKeywordsOption,
      ownProps: {
        searchQuery: {query: otherKeyword},
        setSearch: this.setOtherKeywords,
        clearSearch: this.clearOtherKeywords,
        placeholder: '',
      },
      key: 'other',
      text: localizedKeywordLabels.other,
      value: 'other',
    })

    const Selection = this.props.asMenuList ? MenuList : NavDropdown
    return (
      <Selection placeholder={anyOption.text}
                 label={this.props.intl.formatMessage(messages.keywordLabel)}
                 options={selectOptions}
                 value={_isEmpty(this.props.challengeFilter.keywords) ?
                        null : activeCategory}
                 onChange={this.updateFilter}
    />
    )
  }
}

FilterByKeyword.propTypes = {
  /** Invoked to update the challenge keyword filter */
  setKeywordFilter: PropTypes.func.isRequired,
  /** Invoked to clear the challenge keyword filter */
  removeChallengeFilters: PropTypes.func.isRequired,
  /** The current value of the challenge filter */
  challengeFilter: PropTypes.object,
  /** Set to true to render a MenuList instead of NavDropdown */
  asMenuList: PropTypes.bool,
}

FilterByKeyword.defaultProps = {
  challengeFilter: {},
  asMenuList: false,
}

export default injectIntl(FilterByKeyword)
