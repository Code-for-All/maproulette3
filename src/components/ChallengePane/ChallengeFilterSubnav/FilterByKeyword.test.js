import React from 'react'
import { FilterByKeyword } from './FilterByKeyword'
import { cloneDeep as _cloneDeep } from 'lodash'

const propsFixture = {
  user: {
    id: 11,
    savedChallenges: [],
  },
  challenges: [
    {
      id: 309,
      name: "Challenge 309",
      blurb: "Challenge 309 blurb",
      description: "Challenge 309 description",
      parent: {
        displayName: "foo",
      }
    },
    {
      id: 311,
      name: "Challenge 311",
      blurb: "Challenge 311 blurb",
      description: "Challenge 311 description",
      parent: {
        displayName: "bar",
      }
    }
  ],
  challengeFilter: {
    difficulty: 1,
    keywords: 'Challenge'
  }
}

let basicProps = null

beforeEach(() => {
  basicProps = _cloneDeep(propsFixture)
  basicProps.setKeywordFilter = jest.fn()
  basicProps.removeChallengeFilters = jest.fn()
  basicProps.intl = {formatMessage: jest.fn()}
})

test("it renders with props as expected", () => {
  const wrapper = shallow(
    <FilterByKeyword {...basicProps} />
  )

  expect(wrapper).toMatchSnapshot()
})

test("it calls setKeywordFilter if an onChange occurs with a value", () => {
  const wrapper = mount(
    <FilterByKeyword {...basicProps} />
  )

  wrapper.instance().updateFilter({value: 'water'})
  expect(basicProps.setKeywordFilter).toBeCalledWith(["natural", "water"])
})

test("it calls removeChallengeFilters if an onChange occurs with a null value", () => {
  const wrapper = mount(
    <FilterByKeyword {...basicProps} />
  )

  wrapper.instance().updateFilter({value: null})
  expect(basicProps.removeChallengeFilters).toBeCalledWith(['keywords'])
})
