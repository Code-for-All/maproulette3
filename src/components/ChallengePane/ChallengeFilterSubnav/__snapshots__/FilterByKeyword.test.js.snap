// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders with props as expected 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <FilterByKeyword
    challengeFilter={
        Object {
            "difficulty": 1,
            "keywords": "Challenge",
          }
    }
    challenges={
        Array [
            Object {
              "blurb": "Challenge 309 blurb",
              "description": "Challenge 309 description",
              "id": 309,
              "name": "Challenge 309",
              "parent": Object {
                "displayName": "foo",
              },
            },
            Object {
              "blurb": "Challenge 311 blurb",
              "description": "Challenge 311 description",
              "id": 311,
              "name": "Challenge 311",
              "parent": Object {
                "displayName": "bar",
              },
            },
          ]
    }
    intl={
        Object {
            "formatMessage": [Function],
          }
    }
    removeChallengeFilters={[Function]}
    setKeywordFilter={[Function]}
    user={
        Object {
            "id": 11,
            "savedChallenges": Array [],
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "eventTypes": Array [
        "mousedown",
        "touchstart",
      ],
      "excludeScrollbar": false,
      "label": undefined,
      "onChange": [Function],
      "options": Array [
        Object {
          "key": "any",
          "text": undefined,
          "value": null,
        },
        Object {
          "key": "navigation",
          "text": undefined,
          "value": "navigation",
        },
        Object {
          "key": "water",
          "text": undefined,
          "value": "water",
        },
        Object {
          "key": "pointsOfInterest",
          "text": undefined,
          "value": "pointsOfInterest",
        },
        Object {
          "key": "buildings",
          "text": undefined,
          "value": "buildings",
        },
        Object {
          "key": "landUse",
          "text": undefined,
          "value": "landUse",
        },
        Object {
          "key": "transit",
          "text": undefined,
          "value": "transit",
        },
        Object {
          "Renderable": [Function],
          "key": "other",
          "ownProps": Object {
            "clearSearch": [Function],
            "placeholder": "",
            "searchQuery": Object {
              "query": "C",
            },
            "setSearch": [Function],
          },
          "text": undefined,
          "value": "C",
        },
      ],
      "outsideClickIgnoreClass": "ignore-react-onclickoutside",
      "placeholder": undefined,
      "preventDefault": false,
      "stopPropagation": false,
      "value": Object {
        "text": undefined,
      },
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "eventTypes": Array [
          "mousedown",
          "touchstart",
        ],
        "excludeScrollbar": false,
        "label": undefined,
        "onChange": [Function],
        "options": Array [
          Object {
            "key": "any",
            "text": undefined,
            "value": null,
          },
          Object {
            "key": "navigation",
            "text": undefined,
            "value": "navigation",
          },
          Object {
            "key": "water",
            "text": undefined,
            "value": "water",
          },
          Object {
            "key": "pointsOfInterest",
            "text": undefined,
            "value": "pointsOfInterest",
          },
          Object {
            "key": "buildings",
            "text": undefined,
            "value": "buildings",
          },
          Object {
            "key": "landUse",
            "text": undefined,
            "value": "landUse",
          },
          Object {
            "key": "transit",
            "text": undefined,
            "value": "transit",
          },
          Object {
            "Renderable": [Function],
            "key": "other",
            "ownProps": Object {
              "clearSearch": [Function],
              "placeholder": "",
              "searchQuery": Object {
                "query": "C",
              },
              "setSearch": [Function],
            },
            "text": undefined,
            "value": "C",
          },
        ],
        "outsideClickIgnoreClass": "ignore-react-onclickoutside",
        "placeholder": undefined,
        "preventDefault": false,
        "stopPropagation": false,
        "value": Object {
          "text": undefined,
        },
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
